### PR TITLE
KOGITO-4625: Implement work item definition it coverage in VSCode

### PR DESCRIPTION
**Pull request by domhanak**
Rework how we open file to allow expected file structure in sidebar.

Add work item definition test.

Draft PR, testing stability of changes on CI.

### CODE REVIEW

**Comment by domhanak** [comment]
Mac seems to not work with latest openFolder changes. Will try a workaround.

**Comment by domhanak** [comment]
Mac-os failure still happens, looking for workaround

File: packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/extension-editors.test.ts [hunk 1]
```diff
@@ -145,4 +148,41 @@ describe("Editors are loading properly", () => {
 
     await webview.switchBack();
   });
+
+  it("Opens process with work item definition properly", async function() {
+    this.timeout(20000);
+    webview = await testHelper.openFileFromSidebar(WID_BPMN, "resources/src/main/java/org/kie/businessapp");
```
Comment (jomarko) [inline packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/extension-editors.test.ts]: shouldn't we use `path.resolve` ? as we do for `const RESOURCES: string = path.resolve("src", "ui-test", "resources")`?

just asking, not an expert, I would expect the second is more safe on different platforms?

File: packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/extension-editors.test.ts [hunk 2]
```diff
@@ -145,4 +148,41 @@ describe("Editors are loading properly", () => {
...
+    const customTasksPaletteCategory = await bpmnEditorTester.openDiagramPalette(PaletteCategories.CUSTOM_TASKS);
+    assertWebElementIsDisplayedEnabled(await customTasksPaletteCategory.findElement(h5ComponentWithText("Milestone")));
+    assertWebElementIsDisplayedEnabled(await customTasksPaletteCategory.findElement(h5ComponentWithText("CustomTasks")));
+    assertWebElementIsDisplayedEnabled(await customTasksPaletteCategory.findElement(palletteItemAnchor("CreateCustomer")));
+    assertWebElementIsDisplayedEnabled(await customTasksPaletteCategory.findElement(palletteItemAnchor("Email")));
+
+    const explorer = await bpmnEditorTester.openDiagramExplorer();
+    let customTaskAnchor = await explorer.findElement(By.xpath(aComponentWithText("Create Customer Internal Service"))); //store to click after checks.
+    await assertWebElementIsDisplayedEnabled(customTaskAnchor);
+    await assertWebElementIsDisplayedEnabled(await explorer.findElement(By.xpath(aComponentWithText("ProcessWithWorkItemDefinition"))));
```
Comment (jomarko) [inline packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/extension-editors.test.ts]: To shorten codebase, could be worth to return `By` directly from `CommonLocators.ts`, what is your opinion?

**Review by jomarko — COMMENTED** [review]
Not passed on my machine. I did:
- `yarn run build:prod` over `kogito-tooling`
- `yarn run test:it` over `kogito-tooling/packages/vscode-extension-pack-kogito-kie-editors`

Do I miss some step in build please? Attaching test report I got:

```xml
<testcase classname="Editors are loading properly" name="Opens process with work item definition properly" time="12.525"><failure>no such element: Unable to locate element: {&#x22;method&#x22;:&#x22;xpath&#x22;,&#x22;selector&#x22;:&#x22;//a[.//span[text() = &#x27;CreateCustomer&#x27;]]&#x22;}
  (Session info: chrome=87.0.4280.141)
  (Driver info: chromedriver=87.0.4280.88 (89e2380a3e36c3464b5dd1302349b1382549290d-refs/branch-heads/4280@{#1761}),platform=Linux 5.11.14-200.fc33.x86_64 x86_64)
NoSuchElementError: no such element: Unable to locate element: {&#x22;method&#x22;:&#x22;xpath&#x22;,&#x22;selector&#x22;:&#x22;//a[.//span[text() = &#x27;CreateCustomer&#x27;]]&#x22;}
  (Session info: chrome=87.0.4280.141)
  (Driver info: chromedriver=87.0.4280.88 (89e2380a3e36c3464b5dd1302349b1382549290d-refs/branch-heads/4280@{#1761}),platform=Linux 5.11.14-200.fc33.x86_64 x86_64)
    at Object.checkLegacyResponse (/home/jomarko/redhat/github/kiegroup/kogito-tooling/node_modules/vscode-extension-tester/node_modules/selenium-webdriver/lib/error.js:546:15)
    at parseHttpResponse (/home/jomarko/redhat/github/kiegroup/kogito-tooling/node_modules/vscode-extension-tester/node_modules/selenium-webdriver/lib/http.js:509:13)
    at /home/jomarko/redhat/github/kiegroup/kogito-tooling/node_modules/vscode-extension-tester/node_modules/selenium-webdriver/lib/http.js:441:30
    at runMicrotasks (&#x3C;anonymous&#x3E;)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
From: Task: WebElement.findElement(By(xpath, //a[.//span[text() = &#x27;CreateCustomer&#x27;]]))
    at Driver.schedule (/home/jomarko/redhat/github/kiegroup/kogito-tooling/node_modules/vscode-extension-tester/node_modules/selenium-webdriver/lib/webdriver.js:807:17)
    at WebElement.schedule_ (/home/jomarko/redhat/github/kiegroup/kogito-tooling/node_modules/vscode-extension-tester/node_modules/selenium-webdriver/lib/webdriver.js:2010:25)
    at WebElement.findElement (/home/jomarko/redhat/github/kiegroup/kogito-tooling/node_modules/vscode-extension-tester/node_modules/selenium-webdriver/lib/webdriver.js:2057:17)
    at Context.&#x3C;anonymous&#x3E; (out/ui-test/extension-editors.test.js:345:73)
    at step (out/ui-test/extension-editors.test.js:48:23)
    at Object.next (out/ui-test/extension-editors.test.js:29:53)
    at fulfilled (out/ui-test/extension-editors.test.js:20:58)
    at runMicrotasks (&#x3C;anonymous&#x3E;)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)</failure></testcase>
<testcase classname="Editors are loading properly" name="&#x22;after each&#x22; hook for &#x22;Opens process with work item definition properly&#x22;" time="0.02"><failure>no such element: Unable to locate element: {&#x22;method&#x22;:&#x22;css selector&#x22;,&#x22;selector&#x22;:&#x22;.monaco-workbench&#x22;}
  (Session info: chrome=87.0.4280.141)
  (Driver info: chromedriver=87.0.4280.88 (89e2380a3e36c3464b5dd1302349b1382549290d-refs/branch-heads/4280@{#1761}),platform=Linux 5.11.14-200.fc33.x86_64 x86_64)
NoSuchElementError: no such element: Unable to locate element: {&#x22;method&#x22;:&#x22;css selector&#x22;,&#x22;selector&#x22;:&#x22;.monaco-workbench&#x22;}
  (Session info: chrome=87.0.4280.141)
  (Driver info: chromedriver=87.0.4280.88 (89e2380a3e36c3464b5dd1302349b1382549290d-refs/branch-heads/4280@{#1761}),platform=Linux 5.11.14-200.fc33.x86_64 x86_64)
    at Object.checkLegacyResponse (/home/jomarko/redhat/github/kiegroup/kogito-tooling/node_modules/vscode-extension-tester/node_modules/selenium-webdriver/lib/error.js:546:15)
    at parseHttpResponse (/home/jomarko/redhat/github/kiegroup/kogito-tooling/node_modules/vscode-extension-tester/node_modules/selenium-webdriver/lib/http.js:509:13)
    at /home/jomarko/redhat/github/kiegroup/kogito-tooling/node_modules/vscode-extension-tester/node_modules/selenium-webdriver/lib/http.js:441:30
    at runMicrotasks (&#x3C;anonymous&#x3E;)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
From: Task: WebDriver.findElement(By(css selector, .monaco-workbench))
    at Driver.schedule (/home/jomarko/redhat/github/kiegroup/kogito-tooling/node_modules/vscode-extension-tester/node_modules/selenium-webdriver/lib/webdriver.js:807:17)
    at Driver.findElement (/home/jomarko/redhat/github/kiegroup/kogito-tooling/node_modules/vscode-extension-tester/node_modules/selenium-webdriver/lib/webdriver.js:1014:17)
    at new AbstractElement (/home/jomarko/redhat/github/kiegroup/kogito-tooling/node_modules/vscode-extension-tester/node_modules/monaco-page-objects/out/components/AbstractElement.js:33:43)
    at new EditorView (/home/jomarko/redhat/github/kiegroup/kogito-tooling/node_modules/vscode-extension-tester/node_modules/monaco-page-objects/out/components/editor/EditorView.js:25:9)
    at VSCodeTestHelper.&#x3C;anonymous&#x3E; (out/ui-test/helpers/VSCodeTestHelper.js:196:50)
    at step (out/ui-test/helpers/VSCodeTestHelper.js:48:23)
    at Object.next (out/ui-test/helpers/VSCodeTestHelper.js:29:53)
    at /home/jomarko/redhat/github/kiegroup/kogito-tooling/packages/vscode-extension-pack-kogito-kie-editors/out/ui-test/helpers/VSCodeTestHelper.js:23:71
    at new Promise (&#x3C;anonymous&#x3E;)
    at __awaiter (out/ui-test/helpers/VSCodeTestHelper.js:19:12)
From: Task: WebElement.findElement(By(css selector, *[id=&#x22;workbench\.parts\.editor&#x22;]))
    at Driver.schedule (/home/jomarko/redhat/github/kiegroup/kogito-tooling/node_modules/vscode-extension-tester/node_modules/selenium-webdriver/lib/webdriver.js:807:17)
    at WebElementPromise.schedule_ (/home/jomarko/redhat/github/kiegroup/kogito-tooling/node_modules/vscode-extension-tester/node_modules/selenium-webdriver/lib/webdriver.js:2010:25)
    at WebElementPromise.findElement (/home/jomarko/redhat/github/kiegroup/kogito-tooling/node_modules/vscode-extension-tester/node_modules/selenium-webdriver/lib/webdriver.js:2057:17)
    at new AbstractElement (/home/jomarko/redhat/github/kiegroup/kogito-tooling/node_modules/vscode-extension-tester/node_modules/monaco-page-objects/out/components/AbstractElement.js:39:48)
    at new EditorView (/home/jomarko/redhat/github/kiegroup/kogito-tooling/node_modules/vscode-extension-tester/node_modules/monaco-page-objects/out/components/editor/EditorView.js:25:9)
    at VSCodeTestHelper.&#x3C;anonymous&#x3E; (out/ui-test/helpers/VSCodeTestHelper.js:196:50)
    at step (out/ui-test/helpers/VSCodeTestHelper.js:48:23)
    at Object.next (out/ui-test/helpers/VSCodeTestHelper.js:29:53)
    at /home/jomarko/redhat/github/kiegroup/kogito-tooling/packages/vscode-extension-pack-kogito-kie-editors/out/ui-test/helpers/VSCodeTestHelper.js:23:71
    at new Promise (&#x3C;anonymous&#x3E;)
From: Task: WebElement.findElements(By(css selector, .editor-group-container))
    at Driver.schedule (/home/jomarko/redhat/github/kiegroup/kogito-tooling/node_modules/vscode-extension-tester/node_modules/selenium-webdriver/lib/webdriver.js:807:17)
    at EditorView.schedule_ (/home/jomarko/redhat/github/kiegroup/kogito-tooling/node_modules/vscode-extension-tester/node_modules/selenium-webdriver/lib/webdriver.js:2010:25)
    at EditorView.findElements (/home/jomarko/redhat/github/kiegroup/kogito-tooling/node_modules/vscode-extension-tester/node_modules/selenium-webdriver/lib/webdriver.js:2081:19)
    at EditorView.&#x3C;anonymous&#x3E; (/home/jomarko/redhat/github/kiegroup/kogito-tooling/node_modules/vscode-extension-tester/node_modules/monaco-page-objects/out/components/editor/EditorView.js:137:41)
    at Generator.next (&#x3C;anonymous&#x3E;)
    at /home/jomarko/redhat/github/kiegroup/kogito-tooling/node_modules/vscode-extension-tester/node_modules/monaco-page-objects/out/components/editor/EditorView.js:8:71
    at new Promise (&#x3C;anonymous&#x3E;)
    at __awaiter (/home/jomarko/redhat/github/kiegroup/kogito-tooling/node_modules/vscode-extension-tester/node_modules/monaco-page-objects/out/components/editor/EditorView.js:4:12)
    at EditorView.getEditorGroups (/home/jomarko/redhat/github/kiegroup/kogito-tooling/node_modules/vscode-extension-tester/node_modules/monaco-page-objects/out/components/editor/EditorView.js:136:16)
    at EditorView.&#x3C;anonymous&#x3E; (/home/jomarko/redhat/github/kiegroup/kogito-tooling/node_modules/vscode-extension-tester/node_modules/monaco-page-objects/out/components/editor/EditorView.js:58:37)</failure></testcase>
```

File: packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/extension-editors.test.ts (hunk 1 above)
Comment (domhanak) [reply packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/extension-editors.test.ts]: Yup, definitely. I will refactor the code before I ask for final review.

File: packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/extension-editors.test.ts (hunk 2 above)
Comment (domhanak) [reply packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/extension-editors.test.ts]: Agreed, will refactor `CommonLocators.ts` to return By all the time.

**Comment by domhanak** [comment]
@jomarko do you have latest editors? e.g. 7.52.Final? `yarn run init`

**Comment by jomarko** [comment]
yes, test passed after `yarn run init` usage.

**Comment by tiagobento** [comment]
There are conflicts on this PR, can you please check?

**Comment by domhanak** [comment]
@tiagobento yea, still wip. Sorry forgot to add the tag.

**Comment by domhanak** [comment]
@jomarko look at the speeeed :D

**Comment by jomarko** [comment]
@domhanak thank you. Yest, it seems to be faster now. Are you still troubleshooting the mac?

**Comment by domhanak** [comment]
For some reason this no longer works after master rebase -_- Investigating.

**Comment by domhanak** [comment]
@jomarko ready for re-review I removed the waiting mechanism as it is not longer working well. Now only adds the test and necessary updates to test helpers.

**Comment by tiagobento** [comment]
@domhanak VS Code tests are still not 100% on Windows, as the build keeps failures randomly, it seems..

**Comment by domhanak** [comment]
@tiagobento This PR does not modify anything that would change the way of waiting so stabilization is not a target here.

File: packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/extension-editors.test.ts [hunk 3]
```diff
@@ -145,4 +153,50 @@ describe("Editors are loading properly", () => {
...
+    let docInput = await propertiesPanel.findElement(customTaskDocumentationTextArea());
+    assertWebElementIsDisplayedEnabled(docInput);
+    assert.equal(
+      await docInput.getAttribute("value"),
+      "Calls internal service that creates the customer in database server."
+    );
+
+    let assignmentsTextBox = await propertiesPanel.findElement(assignmentsTextBoxInput());
+    assert.isTrue(await assignmentsTextBox.isEnabled());
+    assert.equal(await assignmentsTextBox.getAttribute("value"), "7 data inputs, 1 data output");
```
Comment (jomarko) [inline packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/extension-editors.test.ts]: From my point of view we could have a helper method as we three times repeat the same:
- find element
- extract its value attribute
- check the extracted value

However I will not block from merge if you do not wish it.

File: packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/extension-editors.test.ts [hunk 4]
```diff
@@ -145,4 +153,50 @@ describe("Editors are loading properly", () => {
...
+    await webview.switchToFrame();
+    const bpmnEditorTester = new BpmnEditorTestHelper(webview);
+
+    const customTasksPaletteCategory = await bpmnEditorTester.openDiagramPalette(PaletteCategories.CUSTOM_TASKS);
+    assertWebElementIsDisplayedEnabled(await customTasksPaletteCategory.findElement(h5ComponentWithText("Milestone")));
+    assertWebElementIsDisplayedEnabled(
+      await customTasksPaletteCategory.findElement(h5ComponentWithText("CustomTasks"))
+    );
+    assertWebElementIsDisplayedEnabled(
+      await customTasksPaletteCategory.findElement(palletteItemAnchor("CreateCustomer"))
```
Comment (jomarko) [inline packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/extension-editors.test.ts:169]: is there some conf needed to see that node? or just the latest VS Code?

Screenshot from 2021-05-12 14-31-37

File: packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/helpers/BpmnEditorTestHelper.ts [hunk 5]
```diff
@@ -100,4 +113,35 @@ export default class BpmnEditorTestHelper {
...
+        categories[i],
+        "Unexpected palette category on index: [" + i + "]. Expected [" + categories[i] + "], but got [" + title + "]."
+      );
+      if (title == categoryToOpen) {
+        paletteElements[i].click();
+      }
+    }
+
+    return await palette.findElement(By.className("kie-palette-flyout"));
+  };
```
Comment (jomarko) [inline packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/helpers/BpmnEditorTestHelper.ts]: From my point of view. You bring quite safe approach to click the category in palette. 

I am not sure about theoretical corner case, when categoryToOpen is not found. Shouldn't be then the last lines more like

```ts
  public openDiagramPalette = async (categoryToOpen: PaletteCategories): Promise<WebElement> => {
    const palette = await this.getPalette();
    await assertWebElementIsDisplayedEnabled(palette);
    const paletteElements = await palette.findElements(By.xpath("//button[@data-field='categoryIcon' and @title]"));
    assert.lengthOf(paletteElements, 9); // there should always be 9 categories

    let categories = Object.values(PaletteCategories);
    let title: string;
    for (let i = 0; i < paletteElements.length; i++) {
      title = await paletteElements[i].getAttribute("title");
      assert.equal(
        title,
        categories[i],
        "Unexpected palette category on index: [" + i + "]. Expected [" + categories[i] + "], but got [" + title + "]."
      );
      if (title == categoryToOpen) {
        paletteElements[i].click();
        return await palette.findElement(By.className("kie-palette-flyout"));
      }
    }

    throw Error(categoryToOpen + ": not found")
  };
```


just checking, not saying I am correct

File: packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/helpers/CommonLocators.ts [hunk 6]
```diff
@@ -43,6 +43,15 @@ export const h3ComponentWithText = (text: string): string => {
   return `//h3[text() = \'${text}\']`;
 };
 
+/**
+ * Creates a XPATH string that locates <h4> element with specific text.
```
Comment (jomarko) [inline packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/helpers/CommonLocators.ts]: h5

File: packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/helpers/VSCodeTestHelper.ts [hunk 7]
```diff
@@ -69,27 +103,48 @@ export default class VSCodeTestHelper {
...
+      // For some reason openItem() collapses the view it expands so we
+      // click on src to reexpand the tree and click on desired item
+      const srcItem = await this.workspaceSectionView.findItem(this.SRC_ROOT);
+      if (srcItem != undefined) {
+        srcItem.click();
+      }
+      const fileItem = await this.workspaceSectionView.findItem(fileName);
+      if (fileItem != undefined) {
+        fileItem.click();
+      }
```
Comment (jomarko) [inline packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/helpers/VSCodeTestHelper.ts]: Should we use:
```ts
if (srcItem != undefined) {
        await srcItem.click();
      }
      const fileItem = await this.workspaceSectionView.findItem(fileName);
      if (fileItem != undefined) {
        await fileItem.click();
      }
```

? not a TS expert. Just asking.

File: packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/resources/global/default-global.wid [hunk 8]
```diff
@@ -0,0 +1,24 @@
+[
+  [
+    "name" : "CreateCustomer",
+    "version" : "1.0.0",
+    "documentation" : "Calls internal service that creates the customer in database server.",
+    "displayName" : "Create Customer Internal Service",
+    "parameters" : [
+	  "in_customer_id" : new StringDataType(),
+      "in_customer_level_id" : new IntegerDataType(),
```
Comment (jomarko) [inline packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/resources/global/default-global.wid]: minor, maybe we can unify the indentation?

**Review by jomarko — COMMENTED** [review]
Just some questions before approval, nothing serious. Thank you @domhanak  

Locally the new test passed.

File: packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/extension-editors.test.ts (hunk 4 above)
Comment (domhanak) [reply packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/extension-editors.test.ts:169]: In order to see that node like this you need to open the `resources` as folder in separate VSCode instance for example. `global` folder contains this WID so it won't show in this case. This issue is reported.

**Comment by domhanak** [comment]
@jomarko Thanks I ll work on adding your comments.

File: packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/helpers/BpmnEditorTestHelper.ts (hunk 5 above)
Comment (jomarko) [reply packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/helpers/BpmnEditorTestHelper.ts]: maybe also `await paletteElements[i].click();`

File: packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/extension-editors.test.ts (hunk 4 above)
Comment (jomarko) [reply packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/extension-editors.test.ts:169]: yes, that works, thank you

File: packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/helpers/VSCodeTestHelper.ts (hunk 7 above)
Comment (domhanak) [reply packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/helpers/VSCodeTestHelper.ts]: I did not do this one. It looks the same as current one. Could you please clarify?

File: packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/extension-editors.test.ts (hunk 3 above)
Comment (domhanak) [reply packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/extension-editors.test.ts]: I made it so that we have an assert that takes WebElement, attribute, attributeValue. Finding elements is very test dependent so I keep it as is.

**Comment by domhanak** [comment]
@jomarko added improvements, let me know! :)

File: packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/helpers/VSCodeTestHelper.ts (hunk 7 above)
Comment (jomarko) [reply packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/helpers/VSCodeTestHelper.ts]: my proposal add `await` for `srcItem.click()` and `fileItem.click()`

Comment (domhanak) [reply packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/helpers/VSCodeTestHelper.ts]: Ahh yes, I am blind :D

**Review by jomarko — APPROVED** [review]
Thank you

**Comment by domhanak** [comment]
@tiagobento should be good to go if OK from your side.

**Review by tiagobento — APPROVED** [review]
